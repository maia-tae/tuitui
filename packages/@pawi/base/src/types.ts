import { TArg, TBlock, TContext, TValue } from 'pawi'
import { BaseContext } from './context'

// 'types' content is never exported and is just a wrapper to
// simplify typing of modules.

// Init function
export type Context = TContext<BaseContext>
export type Block = TBlock<BaseContext>
// Update function (if alone)
export type Arg = TArg<BaseContext>
export type Value = TValue<BaseContext>
