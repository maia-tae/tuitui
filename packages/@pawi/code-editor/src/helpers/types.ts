export const SCRUBBER_VAR = '$scrub$'

export interface LiteralScrub {
  value: number
  text: string
  ch: number
  line: number
}

export interface CMOptions {
  lucidity: EditorLucidityOptions
}

export interface CMEditor extends CodeMirror.Editor {
  options: CMOptions
}

interface EditorLucidityOptions {
  scrubber: Scrubber
  lock?: string
  noscrub?: boolean
  blockId?: string // to detect change of block selection
  cursorMarkCleared?: boolean
  nosave?: boolean
}

export interface Scrubber extends ScrubCode {
  // Call init on block if value changes
  init?: any
}

export interface ScrubCode {
  js: string
  values: number[]
  literals: LiteralScrub[]
}

export interface CompilerError {
  message: string
  line: number
  ch: number
}

export interface TranspileCallbackArgs {
  js: string
  scrub: ScrubCode
  errors: CompilerError[]
}

export interface TranspileCallback {
  ({ js, scrub, errors }: TranspileCallbackArgs): void
}
