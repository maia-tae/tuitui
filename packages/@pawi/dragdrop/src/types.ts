import { IContext } from 'overmind'
import * as actions from './actions'

export type DragdropComponentType =
  | React.ComponentClass<any>
  | React.FunctionComponent<any>

export interface DragdropDefinitions {
  types: {
    [type: string]: {
      hooks?: Partial<DragdropHooks>
      component: DragdropComponentType
      // Extra props to pass to component on drag. This will not be part of the
      // payload. Used to alter behavior or visual of dragged element for example.
      dragProps?: any
      // Fixed dragging position in component
      anchor?: Position
    }
  }
  // This changes the drag type on drag start.
  dragTransform: Transforms
  // This changes a drag type on drop (hover).
  dropTransform: Transforms
}

export interface DragTransform<
  Context = IContext<any>,
  SourcePayload = any,
  TargetPayload = any
> {
  (ctx: Context, dragData: DragData<SourcePayload>):
    | DragData<TargetPayload>
    | undefined
}

export interface Transforms {
  [sourceType: string]: {
    [targetType: string]: DragTransform
  }
}

export interface DragdropSettings {
  // Dragdrop components where we show the rest of the app.
  dragdrop?: Partial<DragdropDefinitions>
}

export interface DragData<Payload = any> {
  // Anchor in dragged object
  anchor: Position
  // When a drag is happening, this is set with the type of dragged element.
  // Drop zones react to this type by changing style to show if they accept it.
  type: string
  // Any kind of information related to the given type that helps the drop zone
  // know what to do with it.
  payload: Payload
}

export interface DropData<Payload = any> {
  // Type of drop zone (used to transform on release).
  type: string
  // When the drag ends, this is called (set when hovering above a Drop element). The
  // call is async and will behave as if it was called from the Drop target.
  // No need for the extra function wrapping to avoid Derive because type is optional.
  callback: OnDropFunction
  // Information on drop added.
  payload: Payload
}

export interface Position {
  x: number
  y: number
}

// Hooks triggered in dragdrop actions

// UI hooks inserted in html component
export interface DraggableHooks {
  className?: string
  onClick(e: any): any
  // This is for very small draggable items (we cannot start drag before being outside of element).
  onMouseLeave(e: any): any
  onMouseUp(e: any): any
  onMouseDown(e: any): any
  onMouseMove(e: any): any
}

export interface DraggableOptions {
  onClick?: () => void
  // When this flag is set to true, dragging is disabled.
  noDrag?: boolean
}

// UI hooks inserted in html component
export interface DroppableHooks {
  // Detect mouse enter, also on initial drag zone if drag starts there.
  onMouseMove(e: any): any
  onMouseLeave(e: any): any
  className: string
}

export interface FileDroppableHooks {
  // Just to make sure it is inserted in the HTML element.
  ref: React.RefObject<any>
  onDrop(e: any): any
  onDragEnter(e: any): any
  onDragOver(e: any): any
  onDragLeave(e: any): any
}

export interface DroppableSettings<T = any> {
  className?: string
  // Is the drop enabled ? If the value is a `string`, this is used as
  // extra className to explain why it is disabled.
  enable?: boolean | string
  // Type of draggable item to accept. If we want to accept more types, we can
  // create a wrapper type and use dropTransform to transform incoming types to our
  // wrapper type.
  drop: string
  onDrop: (dragPayload: any) => void
  // Extra payload to add to draggable payload (can provide context).
  payload?: T
}

export interface FileDroppableOptions {
  // We need to use a ref as the drag operations do not touch
  // overmind and we need to update the div.
  ref: React.RefObject<HTMLElement>
  className?: string
  onDrop: (args: { files: File[], strings: [] }) => void
  accept?: (arg: { kind: string, type: string}) => boolean
  // Extra payload to add
  payload?: any
}

export interface DragdropState {}

export type OnDropFunction = (dragPayload: any) => void

// Action hooks of each draggable type
export interface DragdropHooks {
  // If 'true' is returned, normal action is ignored (run considered done)
  enter: (ctx: any, value: actions.EnterArg) => boolean | void
  leave: (ctx: any) => boolean | void
  move: (ctx: any, value: actions.MoveArg) => boolean | void
  release: (ctx: any) => boolean | void
  start: (ctx: any, value: actions.StartArg) => boolean | void
}

export interface DragdropConfig {
  state: {
    dragdrop: {
      // This can be set during a drag operation to disable all drop
      // zones.
      disableDrop?: boolean
      // Dragged data
      drag?: DragData
      // Mouse position
      position: Position
      drop?: DropData
      definitions: () => DragdropDefinitions
    }
  }

  actions: {
    dragdrop: typeof actions
  }
}
