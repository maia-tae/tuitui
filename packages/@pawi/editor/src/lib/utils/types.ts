import * as React from 'react'
import { ApplyOpArgs } from '../../actions'
import { Comp } from '../../app'
import { OperationsKey } from '../doOperation'

// ============== Composition definition

export type PathType = string[]

export interface ElementsType {
  [key: string]: ElementType
}

// Used during page preview

export interface ElementSize {
  // Content with padding.
  // content
  c: number
  // top margin
  t: number
  // bottom margin
  b: number
}

export interface LegacySize {
  // Legacy format
  content?: number
  topMargin?: number
  bottomMargin?: number
}

export interface ElementSizes {
  [key: string]: ElementSize
}

export interface PositionedElementType {
  p: number
}

export interface PositionedElementsType {
  [key: string]: PositionedElementType
}

export interface ElementOptionsType {
  // custom type name (used to get other options such as className, etc)
  t?: string
  // heading level
  h?: number
  // column attibution
  u?: 'l' | 'r'
  // column width
  uw?: number
  // list type
  l?: 'ol' | 'ul'
  href?: string
  // When true: export this paragraph.
  e?: boolean
  // When true: new page before this paragraph
  n?: boolean
  // align (justify, left, center, right)
  a?: 'j' | 'l' | 'c' | 'r'
  // When this is set, special behavior apply like
  // - cannot remove
  // - cannot change paragraph
  // - cannot set bold/etc (no toolbox)
  title?: boolean
  // For custom elements: open popup on selection
  open?: boolean
}

export interface ParagraphOptions {
  // Custom tag
  c?: string
  // Options to display paragraphs
  o?: ElementOptionsType
}

export interface GroupElementRefTypeById {
  [key: string]: GroupElementRefType
}

export interface GroupElementRefType {
  path: PathType
  elem: GroupElementType
}

export type Type = 'P' | 'A' | 'E' | 'T' | 'B' | 'I' | 'B+I'

export interface ElementType extends ParagraphOptions {
  // Type
  t: Type
  // Position
  p: number
  // Inner content (string)
  i?: string
  // Selection
  s?: SelectionType
  // Group
  g?: ElementsType
}

export interface StringElementType extends ElementType {
  i: string
}

export interface GroupElementType extends ElementType {
  g: ElementsType
}

export interface CustomElementType extends ElementType {
  c: string
}

export interface TitleElementType extends ElementType {
  o: { h: number }
}

export interface ListElementType extends ElementType {
  o: { l: 'ol' | 'ul' }
}

export function isStringElement(elem: ElementType): elem is StringElementType {
  return typeof elem.i === 'string'
}

export function isGroupElement(elem: ElementType): elem is GroupElementType {
  return typeof elem.g === 'object'
}

export function isCustomElement(elem: ElementType): elem is CustomElementType {
  return typeof elem.c === 'string'
}

export function isDocumentTitle(elem: ElementType): elem is TitleElementType {
  return (elem.o && elem.o.title) || false
}

export function isListItem(elem: ElementType): elem is ListElementType {
  const { o } = elem
  return o ? o.l === 'ol' || o.l === 'ul' : false
}

// Main document title
export function isTitle(elem: ElementType): elem is TitleElementType {
  const { t, o } = elem
  return (t === 'P' && o && o.title && true) || false
}

// Any heading
export function isHeading(elem: ElementType): elem is TitleElementType {
  const { t, o } = elem
  return (t === 'P' && o && o.h && true) || false
}

export const isSpecialElement = isCustomElement

export function isRangeSelection(
  selection: SelectionType
): selection is RangeSelectionType {
  return selection.type === 'Range'
}

export interface ElementRefType {
  path: PathType
  elem: ElementType
}

export interface ChangeType {
  op: 'noop' | 'update' | 'delete' | 'create'
  selected?: boolean
  path: PathType
  pathId: string
  elem: ElementType
  resized?: boolean
}

export interface ElementRefTypeById {
  [key: string]: ElementRefType
}

export interface StringElementRefType {
  path: PathType
  elem: StringElementType
}

export interface ElementNamedType {
  ref: string
  elem: ElementType
}

export interface CompositionType {
  sz?: ElementSizes

  toolbox?: ToolboxOperationValueType
  // Data for custom components.
  data?: { [key: string]: any }
  g: ElementsType
  // Path to object containing selection encoded as 'key.subKey.subsubKey'
  spath?: string
  // selection min paragraph position
  smin?: number
  // selection max paragraph position
  smax?: number
  // Title of the composition (when there is one)
  title?: string
}

// ============== Selections

export interface SelectionPositionType {
  top: number
  left: number
}

export interface CaretSelectionType {
  anchorPath: PathType
  anchorOffset: number
  anchorValue?: string
  position: SelectionPositionType
  stringPath?: string // Only used by setSelection
  type: 'Caret'
}

export interface RangeSelectionType {
  anchorPath: PathType
  anchorOffset: number
  anchorValue?: string
  focusPath: PathType
  focusOffset: number
  position: SelectionPositionType
  stringPath?: string // Only used by setSelection
  type: 'Range'
}

export type SelectionType = CaretSelectionType | RangeSelectionType

export interface ChangesType {
  // Initial data for custom elements (further changes do not go through
  // editor).
  data?: { [elemId: string]: any }
  // Full elements.
  elements: { [elemPath: string]: ChangeType }
  start: ChangeType
  end: ChangeType
  // Special cases where we need to prepare selection/toolbox during op eval.
  selection?: SelectionType
}

export function filterChanges(
  elements: ChangesType['elements'],
  types: ChangeType['op'] | ChangeType['op'][] | 'selected'
): ChangeType[] {
  if (types === 'selected') {
    return Object.keys(elements)
      .map(k => elements[k])
      .filter(el => el.selected)
  }

  const filter =
    typeof types === 'string'
      ? ({ op }: ChangeType) => op === types
      : ({ op }: ChangeType) => types.includes(op)
  return Object.keys(elements)
    .map(k => elements[k])
    .filter(filter)
}

// ============== Operations

export interface SelectOperationType {
  op: 'select'
  value: SelectionType
}

export interface UpdateOperationType {
  op: 'update'
  path: PathType
  value: ElementType
}

export interface UpdateOptsOperationType {
  op: 'updateOpts'
  path: PathType
  opts: ParagraphOptions
}

export interface DataOperationType {
  op: 'data'
  path: PathType
  data: any
}

export interface DeleteOperationType {
  op: 'delete'
  path: PathType
}

export interface ResizedOperationType {
  op: 'resized'
  id: string
}

export interface ToolboxOpType {
  type: 'emptyParagraph' | 'paragraph' | 'selection'
  position: SelectionPositionType
}

interface ToolboxNoneType {
  type: 'none'
}

export type ToolboxOperationValueType = ToolboxOpType | ToolboxNoneType

export interface ToolboxOperationType {
  op: 'toolbox'
  value?: ToolboxOperationValueType
}

export type OperationType =
  | SelectOperationType
  | UpdateOperationType
  | UpdateOptsOperationType
  | DataOperationType
  | DeleteOperationType
  | ResizedOperationType
  | ToolboxOperationType

export type OperationsType = OperationType[]

export interface DoOperationType {
  // Not sure about these definitions
  ref: string
  elem: ElementType
}

// =========== Editor Options

export interface TextMarkup {
  (text: string): React.ReactElement<any> | undefined
}

export interface ParagraphProps<Data = any, CProps = any> {
  // Extra tag props passed from Editor.
  customTagProps: CProps
  id: string
  // paragraph
  paragraph: CustomElementType
  // custom data
  data: Data
  holder: CompositionHolder
}

type CustomTag = Comp<ParagraphProps>

export interface SimpleParagraphOption extends BaseParagraphOption {
  // Current paragraph changes
  o: ElementOptionsType
}

// Internal for now
export interface SelectParagraphOption extends BaseParagraphOption {
  op: 'B' | 'I'
}

interface AnyData {
  [key: string]: any
}

export interface InitData<T = AnyData> {
  data: T
  o?: ElementOptionsType
  // Multiple toolbox entries can create the same type of paragraph
  // so this is to overwrite 'toolbox entry key => c' mapping.
  c?: string
}

interface BaseContext {
  state: any
}

export interface InitFunction<
  Context extends BaseContext = { state: any },
  T = AnyData
> {
  (ctx: { state: Context['state'] }, value: ApplyOpArgs): InitData<T>
}

export type InitParagraph = InitData | InitFunction<any, any>

export interface ShowOn {
  paragraph?: boolean
  emptyParagraph?: boolean
  selection?: boolean
}

export interface BaseParagraphOption {
  // as 'cmd+i', 'cmd+shift+I', etc
  // DOES THIS WORK ?
  accel?: string
  // Icon to display in toolbox and when dragging this paragraph...
  icon?: string
  // Icon to display next to text (read during display using o.t type)
  paragraphIcon?: string
  // Custom class name (read during display using o.t type)
  className?: string
  toolboxComponent?: React.ReactNode
  // operation to trigger ('B', 'I', 'A', 'P', etc)
  op?: OperationsKey
  showOn?: ShowOn
  startText?: string
}

export interface CustomParagraphOption<T = any> extends BaseParagraphOption {
  // Initial state for the paragraph's data (if any). Should
  // be an object.
  init: InitParagraph
  // Component to use for the paragraph.
  component: CustomTag
  // Popup to display when custom paragraph is selected (overlay)
  popup?: CustomTag
  // Global widgets that need to be preloaded. These can be heavy UI elements used in a single
  // custom paragraph at a time that we cannot afford to create for every paragraph.
  preload?: Comp<any>
  // On paste, test custom para and creates element (replaces 'markup')
  markup?: (text: string) => T | undefined
  // Provide a function to return a paragraph info (used in paragraph sharing, extraction, etc)
  getInfo?: (ctx: any, data: any) => { title: string; icon: string } | undefined
  // If true, treat this as a special text paragraph with content editing enabled.
  // (this requires managing edits before they reach the editor event handlers: onSelect, onChange).
  noCursor?: boolean
}

export type ParagraphOption =
  | SimpleParagraphOption
  | CustomParagraphOption
  | SelectParagraphOption

export function isSimpleParagraphOption(
  para: ParagraphOption
): para is SimpleParagraphOption {
  return (para as any).o !== undefined
}

export function isCustomParagraphOption(
  para: ParagraphOption
): para is CustomParagraphOption {
  return (para as any).init !== undefined
}

export function isSelectParagraphOption(
  para: ParagraphOption
): para is SelectParagraphOption {
  return (para as any).op !== undefined
}

export interface Paragraphs {
  [key: string]: ParagraphOption
}

export interface EditorOptions {
  paragraphs?: Paragraphs
  // Set this to true to avoid editor default paragraphs.
  noDefaults?: boolean
  paste?: { [type: string]: PasteOperation }
}

// PRIVATE
export interface ParagraphPayload {
  // Initial state for the paragraph's data (if any).
  c?: string
  // On initial paragraph, 'data' is not set in signal props
  data?: AnyData
  o?: ElementOptionsType
  i?: string
}

export interface InternalParagraphOption<T = any> extends BaseParagraphOption {
  // The paragraph key
  key: string
  op: OperationsKey
  // Custom paragraph key
  c?: string
  // This must be set to true for sub-compositions
  isComposition?: boolean
  // JSON.stringify ( ParagraphPayload )
  payload: string
  preload?: ComponentType<any>
  getInfo?: (ctx: any, data: any) => { title: string; icon: string } | undefined
  init?: InitParagraph
  // Component to use for custom paragraphs.
  component?: ComponentType<ParagraphProps>
  popup?: ComponentType<ParagraphProps>
  markup?: (text: string) => T | undefined
  showOn: ShowOn
  // If true, do not insert cursor input (this means we handle cursor selection).
  noCursor?: boolean
}

type ComponentType<P = any> =
  | React.StatelessComponent<P>
  | React.ComponentClass<P>

export type GetMarkup = (
  text: string
) =>
  | {
      // Information to create new custom paragraph.
      c: string
      data: any
    }
  | undefined

export interface EditorProvider {
  preload: ComponentType<any>[]
  paragraphs: { [key: string]: InternalParagraphOption }
  getMenuChoices(type: string): InternalParagraphOption[]
  getMarkup: GetMarkup
  paragraphList: InternalParagraphOption[]
  paste: {
    [type: string]: PasteOperation
  }
  startText: {
    [startString: string]: InternalParagraphOption // simple or custom para ref
  }
}

// ACTIONS AND COMPONENT ARGS

// This is the type passed to the Editor
export interface CompositionHolder {
  id?: string
  title?: string
  $changed?: boolean
  composition?: CompositionType
}

export interface CompositionWrapper {
  composition: CompositionType
}

export interface CompositionArgs {
  holder: CompositionHolder
}

export interface ProcessOpsArgs extends CompositionArgs {
  ops?: OperationType[] | undefined
}

export interface CompositionCaretSelectionArgs extends CompositionArgs {
  selection: CaretSelectionType
}

export interface CompositionSelectionArgs extends CompositionArgs {
  selection: SelectionType
}

export interface InputChangeArgs extends CompositionArgs {
  selection: SelectionType
  value: string | null
}

// ============== Paste types

export interface PasteArgs {
  holder: CompositionHolder
  selection: SelectionType
  event: React.ClipboardEvent<HTMLDivElement>
}

export interface PasteOperation<C = any> {
  // Should return true if the paste is done and false if the content
  // is not useable.
  (ctx: C, type: string, value: PasteArgs): Promise<boolean>
}

export const PASTE_EDITOR = 'pawi/editor'
export const PASTE_HTML = 'text/html'
export const PASTE_TEXT = 'text/plain'
