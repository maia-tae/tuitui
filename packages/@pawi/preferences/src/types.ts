import { Hook } from '@pawi/hooks'
import * as actions from './actions'
import { hooksActions } from './hooksActions'
export const PREFERENCES_KEY = 'pawi/preferences'
export const preferences_restore = 'preferences_restore'
export const preferences_clear = 'preferences_clear'
export const preferences_restored = 'preferences_restored'
export const preferences_save = 'preferences_save'

// Dot based paths
export interface PreferencesPaths {
  // If not true: do not track.
  // We use an object so that libraries can
  // overwride previous settings with 'false'.
  [path: string]: boolean
}

export interface PreferencesSettings {
  preferences?: {
    paths?: PreferencesPaths
    defaults?: { [path: string]: any }
    // Delay in milliseconds between save operations. Default
    // is 3 seconds (3000)
    throttle?: number
    // App name when storing to localStorage. Defaults to 'lucy'.
    // The storage key is 'appName_preferences'.
    appName?: string
    // App website
    appWebsite?: string
  }
}

export interface ChangedArg {
  changed: { [path: string]: boolean }
}

export interface PreferencesHooks {
  [preferences_clear]?: Hook<undefined>
  [preferences_restore]?: Hook<undefined>
  [preferences_restored]?: Hook<undefined>
  [preferences_save]?: Hook<{ path: string; value: any }>
}

// Remove TS error on action rewrite in setup
const actionsNotReadonly = { ...actions }

export interface PreferencesConfig {
  state: {
    preferences: {
      appName: string
      appWebsite: string

      // PRIVATE
      // List of dot encoded paths that have changed. The latest value
      // will be saved on after throttle timer.
      changed: { [path: string]: boolean }
      // PRIVATE: list of watched paths
      paths: string[]
      // PRIVATE: only exists on app setup (must be set through pref settings)
      defaults: { [path: string]: any }
      restoring?: boolean
    }
  }

  actions: {
    preferences: typeof actionsNotReadonly
    hooks: typeof hooksActions
  }
}
