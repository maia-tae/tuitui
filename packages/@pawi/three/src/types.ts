import { TArg, TBlock, TContext, TValue } from 'pawi'
import { ThreeContext } from './context'

// 'types' content is never exported and is just a wrapper to
// simplify typing of modules.

// Init function
export type Context = TContext<ThreeContext>
export type Block = TBlock<ThreeContext>
// Update function (if alone)
export type Arg = TArg<ThreeContext>
export type Value = TValue<ThreeContext>
