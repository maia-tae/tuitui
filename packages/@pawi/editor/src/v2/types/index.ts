export * from './composition'
export * from './element'
export * from './mutate'
export * from './options'
export * from './position'
export * from './selection'
export * from './toolbox'
export * from './size'
