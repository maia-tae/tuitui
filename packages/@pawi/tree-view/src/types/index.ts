export * from './ComponentType'
export * from './DragDrop'
export * from './GetTextSizeType'
export * from './UILayoutType'
export * from './UINodeType'
export * from './UITreeType'

import { TreeType } from 'pawi'
import * as actions from '../actions'
import { DropTargetInfoById } from './DragDrop'
import { UITreeType } from './UITreeType'

export interface SlotInfo {
  x: number
  y: number
  nodeId: string
  slotIdx: number
}

export interface TreeViewConfig {
  state: {
    treeView: {
      dropTarget: DropTargetInfoById
      libraryTree?: TreeType
      library: { name: string; content: any }[]
      // Cached uimap
      uimap: {
        [key: string]: UITreeType
      }
    }
  }

  actions: {
    treeView: typeof actions
  }
}
