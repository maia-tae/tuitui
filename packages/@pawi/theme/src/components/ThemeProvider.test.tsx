import { build, settings } from '@pawi/build'
import * as React from 'react'
import { expect, it, testRender } from 'test'
import { theme } from '..'
import { defaultTheme } from '../themes'
import { ThemeSettings } from '../types'
import { DirectThemeProvider, ThemeProvider } from './ThemeProvider'

const main = {
  name: 'main',
  settings: settings<ThemeSettings>({
    theme: {
      default: { thisWidth: '12px' },
    },
  }),
}

const app = build(main).using(theme).app()

describe('ThemeProvider', () => {
  it('should render children', () => {
    expect(
      testRender(app, <ThemeProvider scopeName="foo">foobar</ThemeProvider>)
    ).toMatchSnapshot()
  })
})

describe('DirectThemeProvider', () => {
  it('should render with provided theme and scopeName', () => {
    expect(
      testRender(
        <DirectThemeProvider scopeName="bar" theme={defaultTheme}>
          foobar
        </DirectThemeProvider>
      )
    ).toMatchSnapshot()
  })
  it('should render with provided theme and default scopeName', () => {
    expect(
      testRender(
        <DirectThemeProvider theme={defaultTheme}>foobar</DirectThemeProvider>
      )
    ).toMatchSnapshot()
  })
})
