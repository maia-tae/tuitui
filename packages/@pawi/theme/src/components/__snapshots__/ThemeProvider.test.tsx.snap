// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DirectThemeProvider should render with provided theme 1`] = `
<div
  className="theme-bar"
>
  foobar
</div>
`;

exports[`DirectThemeProvider should render with provided theme and default scopeName 1`] = `
<div
  className="theme-app"
>
  foobar
</div>
`;

exports[`DirectThemeProvider should render with provided theme and scopeName 1`] = `
<div
  className="theme-bar"
>
  foobar
</div>
`;

exports[`ThemeProvider should render children 1`] = `
<div
  className="theme-foo"
>
  foobar
</div>
`;
